### PR TITLE
Add v4 authentication and request signing

### PR DESCRIPTION
**Pull request by jbrisbin**

### CODE REVIEW

**Comment by jbrisbin** [comment]
@lehoff @lukebakken Be interested to get any feedback on the separation of concerns so far, as well as request encapsulation via the new `riak_s3_req` module.

I've almost got the v4 auth piece ported over from the old code. I couldn't find it in myself to leave a lot of that code in there as I thought it would be better to strip it down and refactor portions of it to make it easier to read and understand and to make it less fragile and more testable.

File: test/riak_s3_auth_SUITE.erl [hunk 1]
```diff
@@ -40,3 +42,14 @@ v4_header_is_understood(_Config) ->
...
+
+v4_authenticate_is_supported(_) ->
+    S3Req0 = riak_s3_test:make_request([
+        riak_s3_test:amz_content_sha256_header(),
+        riak_s3_test:amz_date_header(),
+        riak_s3_test:auth_header(),
+        riak_s3_test:content_type_header("text/plain")
+    ]),
+    {UserID, AuthHeaders, S3Req} = riak_s3_req:get_authorization(S3Req0),
+    io:format("userid: ~p, props: ~p~n", [UserID, AuthHeaders]),
```
Comment (lukebakken) [inline test/riak_s3_auth_SUITE.erl]: Do you see output when you use `io:format`? I had to use [`pal/2`](http://erlang.org/doc/man/ct.html#pal-2) to see output in the console and logs.

File: src/riak_s3_date_utils.erl [hunk 2]
```diff
@@ -0,0 +1,83 @@
+-module(riak_s3_date_utils).
```
Comment (lukebakken) [inline src/riak_s3_date_utils.erl]: Can [`jam`](https://github.com/basho/jam) be used?

File: test/riak_s3_auth_SUITE.erl (hunk 1 above)
Comment (jbrisbin) [inline test/riak_s3_auth_SUITE.erl]: @lukebakken Inadvertently left that in there. I'm an `io:format` debugger by habit.

FWIW this output shows up on the stdout section in the common_test reports.

**Comment by jbrisbin** [comment]
I think this is ready to merge now so we can continue integrating with the request pipeline.

**Comment by lukebakken** [comment]
👍

**Pull request merged**